src/librustc_unicode/tables.rs

```diff
@@ -16,14 +16,19 @@
 /// that the unicode parts of `CharExt` and `UnicodeStrPrelude` traits are based on.
 pub const UNICODE_VERSION: (u64, u64, u64) = (8, 0, 0);
 
-fn bsearch_range_table(c: char, r: &'static [(char,char)]) -> bool {
+fn bsearch_range_table(c: char, r: &'static [(char, char)]) -> bool {
     use core::cmp::Ordering::{Equal, Less, Greater};
     use core::slice::SliceExt;
-    r.binary_search_by(|&(lo,hi)| {
-        if lo <= c && c <= hi { Equal }
-        else if hi < c { Less }
-        else { Greater }
-    }).is_ok()
+    r.binary_search_by(|&(lo, hi)| {
+         if lo <= c && c <= hi {
```

Comment:
Would it make sense to restructure this function to
```Rust
    if c < lo {
        Greater
    } else if hi < c {
        Less
    } else {
        Equal
    }
```
?

The LLVM IR and the assembly from this version are much simpler and unfortunately we cannot rely on the compiler to optimise this much, as it would need to know that `lo <= high`.

Besides from simpler code, there might also be a performance reason why this would be good, although I believe this should not appear on any benchmark, as we are doing a binary search (which should involve few comparisons).
We are performing a search, so the `Equal` branch will be taken exactly once.
If the comparison function returns `Less` L times and `Greater` G times, the old code performs 2+3*L+3*G comparisons. The branch structure I proposed, instead, should perform L+2*G+2 comparisons.